marquez/client.py

```diff
@@ -0,0 +1,237 @@
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import collections
+import os
+import urllib
+
+import requests
+
+from marquez.constants import JobRunState
+from marquez.utils import (InvalidRequestException,
+                           handle_response, _compose_path)
+
+
+class MarquezClient(object):
+    API_PATH = "api/v1"
+    DEFAULT_TIMEOUT_MILLISECONDS = 5000
+    marquez_host = None
+    namespace = None
+
+    def __init__(self, host=None, port=None, timeout=None):
+        host = os.environ.get('MARQUEZ_HOST') or host
+        port = os.environ.get('MARQUEZ_PORT') or (
+            str(port) if port else port)
+        t = os.environ.get('MARQUEZ_TIMEOUT_MS') or str(timeout)
+        timeout_milliseconds = (int(t) if t.isnumeric() else
+                                self.DEFAULT_TIMEOUT_MILLISECONDS)
+        self.timeout = float(timeout_milliseconds / 1000)
+
+        if not host or not port:
+            msg = ("Please provide host & port or set the proper env "
+                   "vars: MARQUEZ_HOST, MARQUEZ_PORT")
+            raise ConnectionError(msg)
+
+        self.marquez_host = "{0}:{1}/{2}".format(
+            host, port, self.API_PATH)
+
+        self._namespace = None
+
+    @property
+    def namespace(self):
+        if not self._namespace:
+            raise InvalidRequestException("Please set a namespace")
+        return self._namespace
+
+    def set_namespace(self, ns_name, owner=None, description=None):
+        self.create_namespace(ns_name, owner or 'default', description)
+        self._namespace = urllib.parse.quote_plus(ns_name)
+
+    #  Namespaces API
+    @handle_response
+    def get_namespace(self, ns_name, **kwargs):
+        if not ns_name:
+            raise InvalidRequestException("Please provide a namespace")
+        path = "namespaces/{0}"
+        path_args = [ns_name]
+        return self.get_request(path, path_args, **kwargs)
+
+    @handle_response
+    def list_namespaces(self, **kwargs):
+        path = "namespaces"
+        path_args = []
+        return self.get_request(path, path_args, **kwargs)
+
+    @handle_response
+    def create_namespace(self, namespace, namespace_owner,
+                         namespace_description=None, **kwargs):
+        path = "namespaces/{0}"
+        path_args = [namespace]
+        payload = {
+            "owner": namespace_owner,
+        }
+        if namespace_description:
+            payload['description'] = namespace_description
+        return self.put_request(path, path_args, payload, **kwargs)
+
+    #  Datasources API
+    @handle_response
+    def list_datasources(self, limit=None, offset=None, **kwargs):
```

Comment:
Why use `**kwargs`? We should be very explicit with the contract we want to expose to our users.

Reply:
There are over a dozen parameters a user may wish to set for a given request (e.g. `timeout`, `allow_redicts`, `headers`, `cookies`, `proxies`, etc.). We don't want to enumerate all of them.

Reply:
Right, but most of them we don't need.

Reply:
we don't want users of this client to know / directly control the behavior of `requests` (or whatever HTTP client we use) unless our client allows it. in this case, we should only allow direct manipulation of the `timeout`. if there are other things we want to expose, we can do that selectively, but it should be a conscious decision.

Reply:
I removed it and replaced with this `timeout_ms`.